### PR TITLE
Support shared secrets for webhook authentication

### PR DESCRIPTION
**Pull request by lizziegooding**
## Context

Adds support for shared secrets to webhooks so webhook receivers can authenticate requests.

## Summary of changes
- [x] add a new `secret` column to webhooks table where we store secrets
- [x] add the option for users to specify their own secret, otherwise generate a random string
- [x] return the secret on `create`; don't return the secret on all other webhook requests
- [x] add url validation to the user passed url on `create` and `update`
- [x] have webhooks send a presigned url with the hmac sha256 hashed payload and shared secret sent as a query parameter named `signature`

## Next steps
[punt] write tests for the `send` function that sends webhooks
- [ ] support secrets in hecate UI
- [ ] test on staging
- [ ] review
- [ ] merge & release

cc/ @ingalls @miccolis @samely @mattciferri

### CODE REVIEW

**Comment by lizziegooding** [comment]
UI updates: adds a webhook secret form on create
Screen Shot 2019-09-30 at 11 24 36 AM

Shows the user a modal on successful create with the webhook secret Hecate responds with
Screen Shot 2019-09-30 at 11 24 46 AM

File: src/webhooks/mod.rs [hunk 1]
```diff
@@ -1,27 +1,47 @@
...
+// Create alias for HMAC-SHA256
+type HmacSha256 = Hmac<Sha256>;
+
 use crate::{
     worker,
     err::HecateError
 };
 
-#[derive(Serialize, Deserialize, Clone)]
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
```
Comment (mattciferri) [inline src/webhooks/mod.rs:17]: Should we remove Debug before deploying?

File: src/webhooks/mod.rs [hunk 2]
```diff
@@ -196,13 +238,25 @@ pub fn send(conn: &impl postgres::GenericConnection, task: &worker::TaskType) ->
             }
         };
 
-        match client.post(hook.url.as_str())
+        let mut mac = match HmacSha256::new_varkey(hook.secret.unwrap().as_bytes()) {
```
Comment (mattciferri) [inline src/webhooks/mod.rs:241]: Is it worth handling the case where hook.secret is None here? I see that in a secret value is always created in the create function, but what about existing webhooks? Just want to ensure we don't panic in that case.

**Review by mattciferri — COMMENTED** [review]
Just a couple comments but looks good overall!

File: src/webhooks/mod.rs (hunk 1 above)
Comment (lizziegooding) [reply src/webhooks/mod.rs:17]: Per chat, going to leave `Debug` as it makes the `Webhooks` struct inspectable. Noting that we haven't been consistent about adding/not adding `Debug` traits to our structures, though we should revisit this in the future.

File: src/webhooks/mod.rs (hunk 2 above)
Comment (lizziegooding) [reply src/webhooks/mod.rs:241]: The `secret` field in the webhooks table is guaranteed to be a non null value https://github.com/mapbox/Hecate/pull/189/files#diff-b22b92c12d5643a0424ec26c1b8e6262R13, which means I'll need to update all existing webhooks anyway. The reason it's an `Option` is so we can handle generating it for the user on `create` if they don't provide, but all functions that retrieve the webhook from the database are guaranteed to have a value for the secret. There may be a different way to handle this than `unwrap`, but it seems appropriate to panic if for some reason that constraint is violated.

Comment (mattciferri) [reply src/webhooks/mod.rs:241]: Nope, that sounds good to me!

**Review by mattciferri — APPROVED** [review]
🌮

**Pull request merged**